buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java

```diff
@@ -0,0 +1,187 @@
+
+    private List<String> ALLOWED_PLATFORMS = Collections.unmodifiableList(Arrays.asList("linux", "windows", "darwin"));
+
+    @Override
+    public void apply(Project project) {
+
+        // create usesJdk task "method"
+        project.getTasks().all(task ->
+            task.getExtensions().findByType(ExtraPropertiesExtension.class).set("usesJdk", new Closure<Void>(project, task) {
+                public void doCall(String version, String platform) {
+                    if (version == null) {
+                        throw new IllegalArgumentException("version must be specified to usesJdk");
+                    }
+                    if (platform == null) {
+                        throw new IllegalArgumentException("platform must be specified to usesJdk");
+                    }
+                    if (ALLOWED_PLATFORMS.contains(platform) == false) {
+                        throw new IllegalArgumentException("platform must be one of " + ALLOWED_PLATFORMS);
+                    }
+                    ExtraPropertiesExtension extraProperties = task.getExtensions().findByType(ExtraPropertiesExtension.class);
+                    if (extraProperties.has("jdkHome")) {
+                        throw new IllegalArgumentException("jdk version already set for task");
+                    }
+
+                    // ensure a root level jdk download task exists
+                    setupRootJdkDownload(project.getRootProject(), platform, version);
+
+                    // setup a configuration for just this version of the jdk
+                    final ConfigurationContainer configurations = project.getConfigurations();
+                    String configurationName = configName("localjdk", version, platform);
+                    Configuration jdkConfig = configurations.findByName(configurationName);
+                    if (jdkConfig == null) {
+                        jdkConfig = configurations.create(configurationName);
+                    }
+                    task.dependsOn(jdkConfig);
+
+                    // depend on the jdk directory "artifact" from the root project
+                    DependencyHandler dependencies = project.getDependencies();
+                    Map<String, Object> depConfig = new HashMap<>();
+                    depConfig.put("path", ":"); // root project
+                    depConfig.put("configuration", configName("localjdk", version, platform));
+                    dependencies.add(configurationName, dependencies.project(depConfig));
+
+                    // add jdkHome runtime property to the task
+                    Object jdkHomeGetter = new Object() {
+                        @Override
+                        public String toString() {
+                            return configurations.getByName(configurationName).getSingleFile().toString();
+                        }
+                    };
+                    extraProperties.set("jdkHome", jdkHomeGetter);
```

Comment:
Rather than registering this in the `ExtraProeprtiesExtension` I think we should define a custom one, has a reference to configurations and does what the above `jdkHomeGetter` would do for this property. 
The advantage is that we get our own namespace, so when build scripts refer to this they do it with `jdkDownloader.jdkHome` and `jdkDownlaoder.usesJDK` making it more obvious where these come from.

Reply:
Sure, I can try to add an extension.

Reply:
I've moved jdkHome to an extension, but I would like to keep usesJdk as it is, because I don't want the extra level of indirection (I think it makes it less like a dsl).

Reply:
Actually I had to revert that even. Apparently an extension must be an abstract type/interface, which means for jdkHome we would need an extra level of indirection there too. So I've kept it as an ext property for now; I'm happy to discuss cleaning this up as a followup.